File: app/src/main/java/com/example/currencyrate/DetailsActivity.java
package com.example.currencyrate;

import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.ScaleAnimation;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.formatter.ValueFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Random;

public class DetailsActivity extends AppCompatActivity {

    private LineChart chart;
    private TextView tvCurrencyCode, tvCurrencyName, tvCurrentRate, tvTrend, tvUnit;
    private final Random random = new Random();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_details);

        chart = findViewById(R.id.chart);
        tvCurrencyCode = findViewById(R.id.tvCurrencyCode);
        tvCurrencyName = findViewById(R.id.tvCurrencyName);
        tvCurrentRate = findViewById(R.id.tvCurrentRate);
        tvTrend = findViewById(R.id.tvTrend);
        tvUnit = findViewById(R.id.tvUnit);

        String code = getIntent().getStringExtra("CURRENCY_CODE");
        if (code == null) code = "USD";
        
        Currency currency = CurrencyData.getByCode(code);

        tvCurrencyCode.setText(currency.getCode());
        tvCurrencyName.setText(currency.getName());
        tvCurrentRate.setText(String.format(Locale.getDefault(), "%.2f %s", currency.getRate(), "₽"));
        tvUnit.setText(String.format(Locale.getDefault(), "за 1 %s", currency.getCode()));

        findViewById(R.id.btnBack).setOnClickListener(v -> finish());

        setupChart();
        updateRandomData(currency.getRate());
        setupButtonAnimations();
    }

    private void setupChart() {
        chart.getAxisLeft().setTextColor(Color.parseColor("#757575"));
        chart.getAxisLeft().setGridColor(Color.parseColor("#22FFFFFF"));
        chart.getAxisLeft().setDrawAxisLine(false);
        chart.getAxisRight().setEnabled(false);
        
        XAxis xAxis = chart.getXAxis();
        xAxis.setPosition(XAxis.XAxisPosition.BOTTOM);
        xAxis.setTextColor(Color.parseColor("#757575"));
        xAxis.setDrawGridLines(false);
        xAxis.setDrawAxisLine(false);
        xAxis.setValueFormatter(new ValueFormatter() {
            private final String[] days = {"18.3", "19.3", "20.3", "21.3", "22.3", "23.3", "24.3"};
            @Override
            public String getFormattedValue(float value) {
                return days[(int) value % days.length];
            }
        });

        chart.getLegend().setEnabled(false);
        chart.getDescription().setEnabled(false);
    }

    private void updateRandomData(double currentRate) {
        List<Entry> entries = new ArrayList<>();
        float baseRate = (float) currentRate;
        for (int i = 0; i < 7; i++) {
            float val = baseRate + (random.nextFloat() * (baseRate * 0.04f) - (baseRate * 0.02f));
            entries.add(new Entry(i, val));
        }

        LineDataSet dataSet = new LineDataSet(entries, "Rate");
        dataSet.setColor(Color.parseColor("#1E88E5"));
        dataSet.setLineWidth(3f);
        dataSet.setDrawCircles(false);
        dataSet.setMode(LineDataSet.Mode.CUBIC_BEZIER);
        dataSet.setDrawFilled(true);
        Drawable drawable = ContextCompat.getDrawable(this, R.drawable.chart_gradient);
        dataSet.setFillDrawable(drawable);
        dataSet.setDrawValues(false);

        LineData lineData = new LineData(dataSet);
        chart.setData(lineData);
        chart.animateX(1000);

        // Random trend
        boolean up = random.nextBoolean();
        float percent = 0.1f + random.nextFloat() * 1.5f;
        tvTrend.setText(String.format(Locale.getDefault(), "%s %.2f%%", (up ? "↗ +" : "↘ -"), percent));
        tvTrend.setTextColor(up ? Color.parseColor("#00C853") : Color.parseColor("#FF1744"));
    }

    private void setupButtonAnimations() {
        findViewById(R.id.btnRefresh).setOnClickListener(v -> {
            v.animate().rotationBy(360).scaleX(0.8f).scaleY(0.8f).setDuration(300).withEndAction(() -> {
                v.animate().scaleX(1f).scaleY(1f).setDuration(200).start();
                Currency c = CurrencyData.getByCode(tvCurrencyCode.getText().toString());
                updateRandomData(c.getRate());
            }).start();
        });

        View tabs = findViewById(R.id.llTabs);
        if (tabs instanceof android.view.ViewGroup) {
            android.view.ViewGroup vg = (android.view.ViewGroup) tabs;
            for (int i = 0; i < vg.getChildCount(); i++) {
                View tab = vg.getChildAt(i);
                tab.setOnClickListener(v -> {
                    animateClick(v);
                    Currency c = CurrencyData.getByCode(tvCurrencyCode.getText().toString());
                    updateRandomData(c.getRate());
                });
            }
        }
    }

    private void animateClick(View v) {
        ScaleAnimation anim = new ScaleAnimation(1f, 0.95f, 1f, 0.95f, 
            Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
        anim.setDuration(100);
        anim.setRepeatMode(Animation.REVERSE);
        anim.setRepeatCount(1);
        v.startAnimation(anim);
    }
}

File: app/src/main/java/com/example/currencyrate/ui/DetailsActivity.kt
package com.example.currencyrate.ui

import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Bundle
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.currencyrate.R
import com.example.currencyrate.data.local.AppDatabase
import com.example.currencyrate.data.remote.CbrApi
import com.example.currencyrate.data.repository.CurrencyRepository
import com.example.currencyrate.databinding.ActivityDetailsBinding
import com.example.currencyrate.viewmodel.DetailsViewModel
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.formatter.ValueFormatter
import retrofit2.Retrofit
import retrofit2.converter.simplexml.SimpleXmlConverterFactory
import java.util.Locale

class DetailsActivity : AppCompatActivity() {

    private lateinit var binding: ActivityDetailsBinding
    @Volatile
    private var currentHistoryDates: List<String> = emptyList()
    private var selectedDays = 7
    
    private val viewModel: DetailsViewModel by viewModels {
        object : ViewModelProvider.Factory {
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                val database = AppDatabase.getDatabase(applicationContext)
                val retrofit = Retrofit.Builder()
                    .baseUrl("https://www.cbr.ru/")
                    .addConverterFactory(SimpleXmlConverterFactory.create())
                    .build()
                val api = retrofit.create(CbrApi::class.java)
                val repository = CurrencyRepository(api, database.currencyDao())
                return DetailsViewModel(repository) as T
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityDetailsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val code = intent.getStringExtra("CURRENCY_CODE") ?: run {
            finish()
            return
        }
        val name = intent.getStringExtra("CURRENCY_NAME") ?: ""
        val rate = intent.getDoubleExtra("CURRENCY_RATE", 0.0)

        setupUI(code, name, rate)
        observeViewModel()

        viewModel.loadHistory(code, selectedDays)
    }

    private fun setupUI(code: String, name: String, rate: Double) {
        binding.btnBack.setOnClickListener { finish() }
        binding.tvCode.text = code
        binding.tvName.text = name

        val formattedRate = if (rate < 0.01) {
            String.format(Locale.getDefault(), "%.4f", rate)
        } else {
            String.format(Locale.getDefault(), "%.2f", rate)
        }
        binding.tvCurrentValue.text = "$formattedRate ₽"

        binding.tab7Days.setOnClickListener { updateTabs(7) }
        binding.tab30Days.setOnClickListener { updateTabs(30) }

        setupChart(binding.chart)
    }

    private fun updateTabs(days: Int) {
        if (days == selectedDays) return
        selectedDays = days

        binding.chart.clear()
        
        val code = binding.tvCode.text.toString()
        viewModel.loadHistory(code, days)
        
        val activeColor = ContextCompat.getColor(this, R.color.accent_blue)
        val inactiveColor = ContextCompat.getColor(this, R.color.text_secondary)
        
        if (days == 7) {
            binding.tab7Days.apply {
                setBackgroundResource(R.drawable.glass_card_bg)
                backgroundTintList = ColorStateList.valueOf(activeColor)
                setTextColor(Color.WHITE)
            }
            binding.tab30Days.apply {
                background = null
                setTextColor(inactiveColor)
            }
        } else {
            binding.tab30Days.apply {
                setBackgroundResource(R.drawable.glass_card_bg)
                backgroundTintList = ColorStateList.valueOf(activeColor)
                setTextColor(Color.WHITE)
            }
            binding.tab7Days.apply {
                background = null
                setTextColor(inactiveColor)
            }
        }
    }

    private fun observeViewModel() {
        viewModel.history.observe(this) { history ->
            binding.chart.clear()
            
            if (history == null || history.isEmpty()) {
                currentHistoryDates = emptyList()
                return@observe
            }
            
            val newDates = history.map { it.date }
            currentHistoryDates = newDates
            
            val entries = history.mapIndexed { index, entity ->
                Entry(index.toFloat(), entity.rate.toFloat())
            }
            
            val marker = CustomMarkerView(this, R.layout.layout_chart_marker, newDates)
            binding.chart.marker = marker
            
            updateChartData(entries)
        }

        viewModel.stats.observe(this) { stats ->
            if (stats != null) {
                binding.tvMin.text = if (stats.first < 0.01) String.format(Locale.getDefault(), "%.4f", stats.first) else String.format(Locale.getDefault(), "%.2f", stats.first)
                binding.tvMax.text = if (stats.second < 0.01) String.format(Locale.getDefault(), "%.4f", stats.second) else String.format(Locale.getDefault(), "%.2f", stats.second)
            }
        }
    }

    private fun setupChart(chart: LineChart) {
        chart.apply {
            description.isEnabled = false
            setTouchEnabled(true)
            setDrawGridBackground(false)
            setExtraOffsets(10f, 10f, 10f, 20f)
            
            xAxis.apply {
                position = XAxis.XAxisPosition.BOTTOM
                textColor = Color.parseColor("#80FFFFFF")
                setDrawGridLines(false)
                setDrawAxisLine(false)
                granularity = 1f
                valueFormatter = object : ValueFormatter() {
                    override fun getFormattedValue(value: Float): String {
                        val index = value.toInt()
                        val dates = currentHistoryDates
                        if (index >= 0 && index < dates.size) {
                            val rawDate = dates[index]
                            return try {
                                val parts = rawDate.split("-")
                                if (parts.size >= 3) "${parts[2]}.${parts[1]}" else rawDate
                            } catch (e: Exception) {
                                rawDate
                            }
                        }
                        return ""
                    }
                }
            }
            
            axisLeft.apply {
                textColor = Color.parseColor("#80FFFFFF")
                setDrawGridLines(true)
                gridColor = Color.parseColor("#1AFFFFFF")
                setDrawAxisLine(false)
            }
            
            axisRight.isEnabled = false
            legend.isEnabled = false
        }
    }

    private fun updateChartData(entries: List<Entry>) {
        if (entries.isEmpty()) {
            binding.chart.clear()
            return
        }
        val dataSet = LineDataSet(entries, "Rate").apply {
            mode = LineDataSet.Mode.CUBIC_BEZIER
            color = ContextCompat.getColor(this@DetailsActivity, R.color.accent_blue)
            lineWidth = 3f
            setDrawCircles(false)
            setDrawValues(false)
            setDrawHighlightIndicators(true)
            highLightColor = Color.WHITE
            highlightLineWidth = 1f
            
            setDrawFilled(true)
            fillDrawable = ContextCompat.getDrawable(this@DetailsActivity, R.drawable.chart_gradient)
        }

        binding.chart.data = LineData(dataSet)
        binding.chart.invalidate()
        binding.chart.animateX(600)
    }
}
